File: Google Mobile Ads/com/smartadserver/android/library/mediation/google/SASGoogleMobileAdsBannerAdapter.java
package com.smartadserver.android.library.mediation.google;

import android.content.Context;
import android.support.annotation.NonNull;
import android.util.Log;
import android.view.View;

import com.google.android.gms.ads.AdListener;
import com.google.android.gms.ads.AdRequest;
import com.google.android.gms.ads.AdSize;
import com.google.android.gms.ads.AdView;
import com.google.android.gms.ads.doubleclick.PublisherAdRequest;
import com.google.android.gms.ads.doubleclick.PublisherAdView;
import com.smartadserver.android.library.mediation.SASMediationBannerAdapter;
import com.smartadserver.android.library.mediation.SASMediationBannerAdapterListener;


import java.util.Map;

/**
 * Mediation adapter class for Google mobile ads banner format
 */
public class SASGoogleMobileAdsBannerAdapter extends SASGoogleMobileAdsAdapterBase implements SASMediationBannerAdapter {

    // tag for logging purposes
    private static final String TAG = SASGoogleMobileAdsBannerAdapter.class.getSimpleName();

    // Google mobile ads banner view instance
    View adView;


    /**
     * Requests a mediated banner ad asynchronously
     *
     * @param context                the {@link android.content.Context} needed by the mediation SDK to make the ad request
     * @param serverParametersString a String containing all needed parameters (as returned by Smart ad delivery)
     *                               to make the mediation ad call
     * @param clientParameters       additional client-side parameters (user specific, like location)
     * @param bannerAdapterListener  the {@link SASMediationBannerAdapterListener} provided to
     *                               this {@link com.smartadserver.android.library.mediation.SASMediationAdapter} to notify Smart SDK of events occurring
     */
    @Override
    public void requestBannerAd(@NonNull Context context, @NonNull String serverParametersString, @NonNull Map<String, String> clientParameters,
                                @NonNull final SASMediationBannerAdapterListener bannerAdapterListener) {

        String adUnitID = serverParametersString.split("\\|")[1];

        GoogleMobileAds gma = initGoogleMobileAds(context, serverParametersString);
        AdSize adSize = getAdSize(serverParametersString);

        if (GoogleMobileAds.ADMOB == gma) {
            // create google ad request
            AdRequest adRequest = configureAdRequest(context, serverParametersString, clientParameters);

            // Create Google AdView and configure it.
            AdView adMobView = new AdView(context);
            adMobView.setAdUnitId(adUnitID);
            adMobView.setAdSize(adSize);

            AdListener adListener =  createAdListener(bannerAdapterListener, adMobView);

            // set listener on banner
            adMobView.setAdListener(adListener);

            // perform ad request
            adMobView.loadAd(adRequest);

            adView = adMobView;

        } else if (GoogleMobileAds.AD_MANAGER == gma) {
            // create google publisher ad request
            PublisherAdRequest publisherAdRequest = new PublisherAdRequest.Builder().build();

            PublisherAdView adManagerView = new PublisherAdView(context);
            adManagerView.setAdUnitId(adUnitID);
            adManagerView.setAdSizes(adSize);

            AdListener adListener =  createAdListener(bannerAdapterListener, adManagerView);

            // set listener on banner
            adManagerView.setAdListener(adListener);

            // perform ad request
            adManagerView.loadAd(publisherAdRequest);

            adView = adManagerView;
        }
    }

    private AdListener createAdListener(SASMediationBannerAdapterListener bannerAdapterListener, View adView) {
        // create Google banner listener that will intercept ad mob banner events and call appropriate SASMediationBannerAdapterListener counterpart methods
        return new AdListener() {

            public void onAdClosed() {
                Log.d(TAG, "Google mobile ads onAdClosed for banner");
                bannerAdapterListener.onAdClosed();
            }

            public void onAdFailedToLoad(int errorCode) {
                Log.d(TAG, "Google mobile ads onAdFailedToLoad for banner (error code:" + errorCode + ")");
                boolean isNoAd = errorCode == AdRequest.ERROR_CODE_NO_FILL;
                bannerAdapterListener.adRequestFailed("Google mobile ads ad loading error code " + errorCode, isNoAd);
            }

            public void onAdLeftApplication() {
                Log.d(TAG, "Google mobile ads onAdLeftApplication for banner");
                bannerAdapterListener.onAdClicked();
                bannerAdapterListener.onAdLeftApplication();
            }

            public void onAdOpened() {
                Log.d(TAG, "Google mobile ads onAdOpened for banner");
            }

            public void onAdLoaded() {
                Log.d(TAG, "Google mobile ads onAdLoaded for banner");
                bannerAdapterListener.onBannerLoaded(adView);
            }
        };

    }


    /**
     * Utility method to get Banner Size from serverParametersString
     */
    protected AdSize getAdSize(String serverParametersString) {
        String[] parameters = serverParametersString.split("\\|");
        int bannerSizeIndex = 0;
        if (parameters.length > 2) {
            // Extracting banner size
            bannerSizeIndex = Integer.parseInt(parameters[2]);
        }
        switch (bannerSizeIndex) {
            case 1:
                return AdSize.MEDIUM_RECTANGLE;
            case 2:
                return AdSize.LEADERBOARD;
            case 3:
                return AdSize.LARGE_BANNER;
        }
        return AdSize.BANNER;
    }




    @Override
    public void onDestroy() {
        Log.d(TAG, "Google mobile ads onDestroy for banner");
        if (adView != null) {
            if (adView instanceof AdView) {
                ((AdView) adView).destroy();
            } else if (adView instanceof  PublisherAdView) {
                ((PublisherAdView)adView).destroy();
            }
        }
    }
}

File: Google Mobile Ads/com/smartadserver/android/library/mediation/google/SASGoogleMobileAdsBannerAdapter.kt
package com.smartadserver.android.library.mediation.google

import android.content.Context
import android.util.Log
import android.view.View
import com.google.android.gms.ads.*
import com.google.android.gms.ads.admanager.AdManagerAdRequest
import com.google.android.gms.ads.admanager.AdManagerAdView
import com.smartadserver.android.library.mediation.SASMediationBannerAdapter
import com.smartadserver.android.library.mediation.SASMediationBannerAdapterListener

/**
 * Mediation adapter class for Google mobile ads banner format
 */
class SASGoogleMobileAdsBannerAdapter : SASGoogleMobileAdsAdapterBase(), SASMediationBannerAdapter {
    // Google mobile ads banner view instance
    var adView: View? = null

    /**
     * Requests a mediated banner ad asynchronously
     *
     * @param context                the [android.content.Context] needed by the mediation SDK to make the ad request
     * @param serverParametersString a String containing all needed parameters (as returned by Smart ad delivery)
     * to make the mediation ad call
     * @param clientParameters       additional client-side parameters (user specific, like location)
     * @param bannerAdapterListener  the [SASMediationBannerAdapterListener] provided to
     * this [com.smartadserver.android.library.mediation.SASMediationAdapter] to notify Smart SDK of events occurring
     */
    override fun requestBannerAd(context: Context,
                                 serverParametersString: String,
                                 clientParameters: Map<String, Any>,
                                 bannerAdapterListener: SASMediationBannerAdapterListener) {
        val adUnitID = getAdUnitID(serverParametersString)
        val gma = initGoogleMobileAds(context, serverParametersString)
        val adSize = getAdSize(serverParametersString)
        if (GoogleMobileAds.ADMOB == gma) {
            // create google ad request
            val adRequest = AdRequest.Builder().build()

            // Create Google AdView and configure it.
            val adMobView = AdView(context)
            adMobView.adUnitId = adUnitID
            adMobView.setAdSize(adSize)
            val adListener = createAdListener(bannerAdapterListener, adMobView)

            // set listener on banner
            adMobView.adListener = adListener

            // perform ad request
            adMobView.loadAd(adRequest)
            adView = adMobView
        } else if (GoogleMobileAds.AD_MANAGER == gma) {
            // create google publisher ad request
            val publisherAdRequest = AdManagerAdRequest.Builder().build()
            val adManagerView = AdManagerAdView(context)
            adManagerView.adUnitId = adUnitID
            adManagerView.setAdSizes(adSize)
            val adListener = createAdListener(bannerAdapterListener, adManagerView)

            // set listener on banner
            adManagerView.adListener = adListener

            // perform ad request
            adManagerView.loadAd(publisherAdRequest)
            adView = adManagerView
        }
    }

    private fun createAdListener(bannerAdapterListener: SASMediationBannerAdapterListener, adView: View): AdListener {
        // create Google banner listener that will intercept ad mob banner events and call appropriate SASMediationBannerAdapterListener counterpart methods
        return object : AdListener() {
            override fun onAdClosed() {
                Log.d(TAG, "Google mobile ads onAdClosed for banner")
                bannerAdapterListener.onAdClosed()
            }

            override fun onAdFailedToLoad(loadAdError: LoadAdError) {
                Log.d(TAG, "Google mobile ads onAdFailedToLoad for banner (error : $loadAdError)")
                val isNoAd = loadAdError.code == AdRequest.ERROR_CODE_NO_FILL
                bannerAdapterListener.adRequestFailed("Google mobile ads banner ad loading error : $loadAdError", isNoAd)
            }

            override fun onAdClicked() {
                Log.d(TAG, "Google mobile ads onAdClicked for banner")
            }

            override fun onAdImpression() {
                Log.d(TAG, "Google mobile ads onAdImpression for banner")
            }

            override fun onAdOpened() {
                Log.d(TAG, "Google mobile ads onAdOpened for banner")
                bannerAdapterListener.onAdClicked()
            }

            override fun onAdLoaded() {
                Log.d(TAG, "Google mobile ads onAdLoaded for banner")
                bannerAdapterListener.onBannerLoaded(adView)
            }
        }
    }

    /**
     * Utility method to get Banner Size from serverParametersString
     */
    private fun getAdSize(serverParametersString: String)= when (serverParametersString.split("|").getOrElse(2){"0"}) {
            "1" -> AdSize.MEDIUM_RECTANGLE
            "2" -> AdSize.LEADERBOARD
            "3" -> AdSize.LARGE_BANNER
            else -> AdSize.BANNER
        }

    override fun onDestroy() {
        Log.d(TAG, "Google mobile ads onDestroy for banner")
        (adView as? AdView)?.destroy() ?: (adView as? AdManagerAdView)?.destroy()
    }

    companion object {
        // tag for logging purposes
        private val TAG = SASGoogleMobileAdsBannerAdapter::class.java.simpleName
    }
}